src/base/EIP712.sol

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IEIP712} from "../interfaces/IEIP712.sol";
+
+/// @notice Generic EIP712 implementation
+/// @dev Maintains cross-chain replay protection in the event of a fork
+/// @dev Should not be delegatecall'd because DOMAIN_SEPARATOR returns the cached hash and does not recompute with the delegatecallers address
+/// @dev Reference: https://github.com/Uniswap/permit2/blob/main/src/EIP712.sol
+/// @dev Reference: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/EIP712.sol
+contract EIP712 is IEIP712 {
+    // Cache the domain separator as an immutable value, but also store the chain id that it
+    // corresponds to, in order to invalidate the cached domain separator if the chain id changes.
+    bytes32 private immutable _CACHED_DOMAIN_SEPARATOR;
+    uint256 private immutable _CACHED_CHAIN_ID;
+    bytes32 private immutable _HASHED_NAME;
+
+    bytes32 private constant _TYPE_HASH =
+        keccak256("EIP712Domain(string name,uint256 chainId,address verifyingContract)");
```

Comment:
why dont we want a version string? i thought part of the idea was that we might have multiple versions? @snreynolds

Reply:
it shaves a bit of gas and sara thinks we can version in name

Reply:
it shouldnt shave anything off gas except in the constructor? after that its not calculating the typehash right?

Reply:
oh you should hardcode this hash - solidity calculates constants at runtime not compile time so itll be recalculating every time